### ISSUE TITLE
Support complex types

### ISSUE THREAD

**Issue by unnonouno** [chainer/chainer#4325]
Currently Chainer assumes integers and floating point numbers. Now CuPy supports complex numbers. It's time to support complex numbers in Chainer.

- [ ] Support complex numbers in gradient check
- [ ] Support complex numbers in various functions

**Comment by m3at** [comment]
I'm curious about this, do you already have use-cases in mind or is that to broaden the capabilities of Chainer?
I see your related [ongoing PR](https://github.com/chainer/chainer/pull/4241), does that target the reproduction of a specific paper?

**Comment by unnonouno** [comment]
Yes. Some research papers use complex numbers or FFT. And i think it can be applied for speech processing. I want to improve capability of Chainer.
- https://arxiv.org/abs/1511.06464
- https://arxiv.org/abs/1611.00035
- http://www.aclweb.org/anthology/D16-1044

**Comment by m3at** [comment]
Understood.
Thanks for the papers, I didn't know about the last two, seems very interesting!
Looking forward to use it in Chainer 👍

**Comment by hknerdgn** [comment]
You should definitely look into https://arxiv.org/pdf/1701.00392.pdf as well.

**Comment by unnonouno** [comment]
Thank you for a link!! Please thumb up this issue if you guys need this feature.

**Comment by toslunar** [comment]
Should Chainer support "complex32" whose real/imag part is `float16`?  It seems NumPy hasn't yet support "complex32".

**Comment by unnonouno** [comment]
IMO, no, especially for the first implementation.

**Comment by andy6975** [comment]
I would like to work on this. Could anyone guide me with the process? Thanks!
